### PR TITLE
test: Creates AD Domain for GMSA e2e testing

### PR DESCRIPTION
**Pull request by JeremyWx**
**Reason for Change**:

Enables automated testing of GMSA feature.

**Requirements**:

- [X ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [X ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:
This extension enables automated e2e testing of the GMSA feature usable by Windows nodes.  It is recommended to run this extension against a pool of one server.  The extension will create the Active Directory domain, the Domain Controller, and the Group Managed Service Account for testing.

### CODE REVIEW

File: extensions/gmsa-dc/v1/GenerateCredentialSpecResource.ps1 [hunk 1]
```diff
@@ -0,0 +1,66 @@
+<#
```
Comment (wk8) [inline extensions/gmsa-dc/v1/GenerateCredentialSpecResource.ps1]: General comment on this file: could be better to port the changes you've made to https://github.com/kubernetes-sigs/windows-gmsa/blob/master/scripts/GenerateCredentialSpecResource.ps1 and then just pull it from there; will help keep everything in sync and less work to maintain it all, what do you think?

File: extensions/gmsa-dc/v1/Promote-DC.ps1 [hunk 2]
```diff
@@ -0,0 +1,51 @@
+###################################################################################
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Promote-DC.ps1]: General nit on this script and all the ones below: would be good to add the usual
```
Set-StrictMode -Version Latest
$ErrorActionPreference = 'Stop'
$PSDefaultParameterValues['*:ErrorAction'] = 'Stop'
```
to exit early on errors - easier to catch and debug issues.

File: extensions/gmsa-dc/v1/Promote-DC.ps1 [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+# This is only for automated e2e testing.  DO NOT use this for production.
+# Jeremy Wood (JeremyWx)
+# Version: 1.0.0.0
+###################################################################################
+
+# Create a working directory and change to it
+mkdir C:\gmsa
+Set-Location -Path C:\gmsa
+
+# Uncomment line below for troubleshooting
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Promote-DC.ps1]: Let's remove this, or actually comment the line below?

File: extensions/gmsa-dc/v1/Promote-DC.ps1 [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+Start-Transcript -Path "C:\gmsa\Promote.txt"
+
+# Download gMSA Setup Script
+Invoke-WebRequest -UseBasicParsing https://raw.githubusercontent.com/JeremyWx/aks-engine/master/extensions/gmsa-dc/v1/Setup-gMSA.ps1 -OutFile Setup-gMSA.ps1
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Promote-DC.ps1]: Point to this repo instead of the fork?

File: extensions/gmsa-dc/v1/Promote-DC.ps1 [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+#Install NuGet
+Install-PackageProvider -Name NuGet -Force
+
+#Create Local Administrator
+$admpassword = ( "K8s" + -join ((48..57) + (97..122) | Get-Random -Count 64 | ForEach-Object {[char]$_}) )
+$admpassword_secure = ( $admpassword | ConvertTo-SecureString -AsPlainText -Force)
+New-LocalUser -Name gmsa-admin -Password $admpassword_secure
+Add-LocalGroupMember -Group "Administrators" -Member gmsa-admin
+Start-Sleep -Seconds 5
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Promote-DC.ps1]: Why is this needed?

File: extensions/gmsa-dc/v1/Promote-DC.ps1 [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+# Make Setup-gMSA run on next boot
+$Logon = "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon"
+$RunOnce = "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\RunOnce"
+Set-ItemProperty $Logon "AutoAdminLogon" -Value "1" -type String
+Set-ItemProperty $Logon "AutoLogonCount" -Value "2" -type DWord
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Promote-DC.ps1]: Why 2? Not sure I understand this line

File: extensions/gmsa-dc/v1/Setup-gMSA.ps1 [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+# Jeremy Wood (JeremyWx)
+# Version: 1.0.0.0
+##################################################################################
+
+# Change to working directory
+Set-Location -Path C:\gmsa
+
+# Script Logging - Feel free to comment out if you like.
+Start-Transcript -Path "C:\gmsa\Setup-gmsa.txt" -Append
+# Making sure the AD services are up and running
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Setup-gMSA.ps1]: Wouldn't there a better way to detect that?

File: extensions/gmsa-dc/v1/Setup-gMSA.ps1 [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Start-Sleep -Seconds 60
+Import-Module ActiveDirectory
+
+if (Get-AdGroupMember -Identity "Enterprise Admins" | Select-String -Pattern "gmsa-admin" -Quiet) {
+    
+	# Add the KDS Root Key
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Setup-gMSA.ps1]: Indentation

File: extensions/gmsa-dc/v1/Setup-gMSA.ps1 [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    $KdsRootKey = Get-KdsRootKey
+    if ($null -eq $KdsRootKey.KeyId) {
+        Add-KdsRootKey –EffectiveTime ((get-date).addhours(-10))
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Setup-gMSA.ps1]: A comment explaining this would be nice to future maintainers :)

(I remember reading something about it being the max propagation delay throughout a cluster, or something?)

File: extensions/gmsa-dc/v1/Setup-gMSA.ps1 [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        Start-Sleep -Seconds 15
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Setup-gMSA.ps1]: Worth a comment too? :)

File: extensions/gmsa-dc/v1/Setup-gMSA.ps1 [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    
+    # Directory for credspecs if not already created
+    mkdir -Path C:\ProgramData\docker\credentialspecs
+
+    # Get GenerateCredSpecResource PowerShell Module
+    #Invoke-WebRequest -UseBasicParsing https://raw.githubusercontent.com/kubernetes-sigs/windows-gmsa/master/scripts/GenerateCredentialSpecResource.ps1 -OutFile GenerateCredentialSpecResource.ps1
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Setup-gMSA.ps1]: Would be better :)

File: extensions/gmsa-dc/v1/Setup-gMSA.ps1 [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    # Run GenerateCredSpecResource to provide the Yaml CredSpec
+    C:\gmsa\GenerateCredentialSpecResource.ps1 gmsa-e2e gmsa-e2e
+
+} else {
+    # Add new admin account to needed groups
+    Add-ADGroupMember -Identity "Enterprise Admins" -Members gmsa-admin
+    Add-ADGroupMember -Identity "Domain Admins" -Members gmsa-admin
+    $RunOnce = "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\RunOnce"
+    Set-ItemProperty $RunOnce "gmsa" -Value "C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe -command C:\gmsa\Setup-gMSA.ps1" -Type String
+    Restart-Computer -Force
```
Comment (wk8) [inline extensions/gmsa-dc/v1/Setup-gMSA.ps1]: Why is this needed?

File: extensions/gmsa-dc/v1/GenerateCredentialSpecResource.ps1 (hunk 1 above)
Comment (JeremyWx) [reply extensions/gmsa-dc/v1/GenerateCredentialSpecResource.ps1]: I have made a PR against the original script and changed the referencing script to pull from the repo location you have suggested above.

https://github.com/kubernetes-sigs/windows-gmsa/pull/16

File: extensions/gmsa-dc/v1/Promote-DC.ps1 (hunk 3 above)
Comment (JeremyWx) [reply extensions/gmsa-dc/v1/Promote-DC.ps1]: I changed the comment to:  # Logging for troubleshooting

Will that work?

File: extensions/gmsa-dc/v1/Promote-DC.ps1 (hunk 4 above)
Comment (JeremyWx) [reply extensions/gmsa-dc/v1/Promote-DC.ps1]: Good catch.  I will change it after I test the other changes you have suggested.

File: extensions/gmsa-dc/v1/Promote-DC.ps1 (hunk 6 above)
Comment (JeremyWx) [reply extensions/gmsa-dc/v1/Promote-DC.ps1]: The process requires 2 reboots.  First we create an admin user and add it to the local Administrators group.  We add some required roles to the server and then reboot to login as our new admin user on our new domain.  Next we add our admin user to the Enterprise Admins group so we can create the KDS root key.  Because group membership is added to your token during login, we need to reboot again so the DC will recognize the admin user as a member of the Enterprise Admins group.

File: extensions/gmsa-dc/v1/Promote-DC.ps1 (hunk 2 above)
Comment (JeremyWx) [reply extensions/gmsa-dc/v1/Promote-DC.ps1]: Thank you for the feedback.  These have been added.

File: extensions/gmsa-dc/v1/Promote-DC.ps1 (hunk 5 above)
Comment (JeremyWx) [reply extensions/gmsa-dc/v1/Promote-DC.ps1]: Honestly, I don't remember why that was added.  It has been removed.

**Comment by JeremyWx** [comment]
Moving to a different repo at request of @PatrickLang
